app/guided.py

```diff
@@ -0,0 +1,388 @@
+                                                                                goal_links)
+            goal_links = await self._get_goal_links(absolute_distance_table, agent=agent)
+
+        if not len(goal_links) or not len(planner_goals):
+            self.planning_svc.log.debug('No more links available or all goals satisfied, planner complete.')
+            self.next_bucket = None
+
+    """ PRIVATE """
+
+    async def _build_attack_graph(self, abilities, agent):
+        """
+        Produces a directed graph that includes nodes for each ability, and nodes of each input and output fact
+        of those abilities.
+        """
+        attack_graph = nx.DiGraph()
+        for ability_id in abilities:
+            for ability in await self.data_svc.locate('abilities', dict(ability_id=ability_id)):
+                if agent:
+                    attack_graph = await self._output_fact_edges(ability, attack_graph, agent)
+                    attack_graph = await self._input_fact_edges(ability, attack_graph, agent)
+                else:
+                    for operation_agent in self.operation.agents:
+                        attack_graph = await self._output_fact_edges(ability, attack_graph, operation_agent)
+                        attack_graph = await self._input_fact_edges(ability, attack_graph, operation_agent)
+        return attack_graph
+
+    async def _output_fact_edges(self, ability, graph, agent):
+        """
+        Adds output facts to a directed graph based on the parserconfigs included in an ability.
+        """
+        executor = await agent.get_preferred_executor(ability)
+        for parser in executor.parsers:
+            for parserconfig in parser.parserconfigs:
+                graph.add_edge(ability, parserconfig.source)
+                if parserconfig.target:
+                    graph.add_edge(ability, parserconfig.target)
+                if parserconfig.edge:
+                    graph.add_edge(ability, parserconfig.edge)
+        return graph
+
+    async def _input_fact_edges(self, ability, graph, agent):
+        """
+        Adds input facts to a directed graph based on the facts present in a command template.
+        """
+        executor = await agent.get_preferred_executor(ability)
+        for fact in re.findall(FACT_REGEX, executor.test, flags=re.DOTALL):
+            if fact not in list(Agent.RESERVED):
+                graph.add_edge(fact, ability)
+        return graph
+
+    async def _create_terminal_goals(self, attack_graph):
+        """
+        Infers goals based on which nodes in a directed graph have an outward degreee of zero.
+        """
+        terminal_nodes = [node for node, degree in attack_graph.out_degree() if degree == 0]
+        aggregated_goals = collections.Counter(terminal_nodes)
+        goals = [Goal(target=trait, operator='*', count=count) for trait, count in aggregated_goals.items()]
```

Comment:
In this case, the graph consists of both abilities and fact traits as nodes. In many cases, nodes with an out degree of zero are actually ability nodes. This causes what is called `trait` here to sometimes be an Ability object if that ability has no output facts. This does not introduce any issues with the code as currently written, but it should be noted that this is not proper to how goals should actually be defined. It would be preferable that all abilities output at least one fact so that ability nodes never have an out degree of zero.